src/linear_algebra/matrix/is_diag.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import linear_algebra.matrix.orthogonal
+import data.matrix.kronecker
+
+/-!
+# Diagonal matrices
+
+This file contains the definition and basic results about diagonal matrices.
+
+## Main results
+
+- `matrix.is_diag`: a proposition that stats a given square matrix `A` is diagonal
+                    (i.e. `∀ m n, i ≠ j → A m n = 0`).
+
+## Tags
+
+diag, diagonal, matrix
+-/
+
+namespace matrix
+
+variables {α β R n m : Type*}
+
+open function
+open_locale matrix kronecker
+
+/-- `A.is_diag` means square matrix `A` is a dianogal matrix: `∀ i j, i ≠ j → A i j = 0`. -/
+def is_diag [has_zero α] (A : matrix n n α) : Prop := ∀ ⦃i j⦄, i ≠ j → A i j = 0
+
+/-- Matrix `A` is diagonal iff there is a vector `d`
+    such that A is the diagonal matix generated by `d`. -/
+lemma is_diag_iff_eq_diagonal [has_zero α] [decidable_eq n] (A : matrix n n α) :
+  A.is_diag ↔ (∃ d, A = diagonal d) :=
+begin
+  split,
+  { intros h,
+    use (λ i, A i i),
+    ext i j,
+    by_cases g : i = j;
+    simp [@h i j, g] },
+  { rintros ⟨d, rfl⟩ i j h, simp [h] }
+end
+
+/-- Every unit matrix is diagonal. -/
+@[simp] lemma is_diag_unit [has_zero α] [subsingleton n] (A : matrix n n α) : A.is_diag :=
```

Comment:
```suggestion
lemma is_diag_of_subsingleton [has_zero α] [subsingleton n] (A : matrix n n α) : A.is_diag :=
```

Reply:
Also, why did you delete the tag here? @jcommelin

Reply:
It's debatable. It's a matter of probabilities. Suppose someone wants to prove `A.is_diag` for some matrix `A`. (Here `A` could be some complicated expression.) The user decides to call `simp`. What is the chance that the indexing type is a subsingleton. Should Lean fire a TC search to find an instance of `[subsingleton n]`? In 99/100 cases, that's wasted time. So I would prefer if the user just calls `is_diag_of_subsingleton` by hand.
But we can keep it in.

Personally, I prefer to reserve `@[simp]` for goals of the form `x = y` or `p \iff q`. That's what simp is really good at.